File: src/evidence/service.ts
import { execFileSync } from 'node:child_process';
import type { CodeGraphNode } from '../codegraph/reader.js';
import { collectGitEvidence, collectGitHubEvidence } from './collector.js';
import type { GitEvidenceCollector } from './git.js';
import type { GitHubEvidenceCollector } from './github.js';
import type { EvidenceRecord, EvidenceStore } from './store.js';

export interface CollectionResult {
  evidence: EvidenceRecord[];
  bundleHash: string;
  source: 'cache' | 'collected';
  collectedAt: number;
  headAtCollection: string | null;
}

export interface EvidenceServiceOptions {
  ttlMs: number;
}

// Single entry point for "give me current evidence for this node".
// Collects on demand, caches with a freshness check that combines a TTL
// with a per-file HEAD-sha comparison: if the file has new commits since
// the last collection, the cache is considered stale even if still
// within TTL. `force: true` bypasses both checks.
export class EvidenceService {
  private readonly ttlMs: number;

  constructor(
    private readonly evidenceStore: EvidenceStore,
    private readonly git: GitEvidenceCollector,
    private readonly github: GitHubEvidenceCollector | null,
    private readonly repoRoot: string,
    opts: EvidenceServiceOptions
  ) {
    this.ttlMs = opts.ttlMs;
  }

  forNode(node: CodeGraphNode, opts?: { force?: boolean }): CollectionResult {
    if (!opts?.force) {
      const cached = this.checkCache(node);
      if (cached) return cached;
    }
    return this.collect(node);
  }

  private checkCache(node: CodeGraphNode): CollectionResult | null {
    const meta = this.evidenceStore.bundleMetaFor(node.id);
    if (!meta) return null;

    const ageMs = Date.now() - meta.built_at;
    if (ageMs > this.ttlMs) return null;

    if (meta.head_at_collection !== null) {
      const currentHead = this.fileHeadSha(node.file_path);
      if (currentHead === null || currentHead !== meta.head_at_collection) {
        return null;
      }
    }
    // If head_at_collection is null (no git history at collection time)
    // we trust the TTL alone — recollecting on every call would be wasteful.

    return {
      evidence: this.evidenceStore.forNode(node.id),
      bundleHash: meta.bundle_hash,
      source: 'cache',
      collectedAt: meta.built_at,
      headAtCollection: meta.head_at_collection,
    };
  }

  private collect(node: CodeGraphNode): CollectionResult {
    const gitRows = collectGitEvidence(this.git, node);
    const ghRows =
      this.github && this.github.isAvailable()
        ? collectGitHubEvidence(this.github, gitRows)
        : [];
    const rows = gitRows.concat(ghRows);
    const headAtCollection = this.fileHeadSha(node.file_path);
    const bundleHash = this.evidenceStore.replace(
      node.id,
      node.qualified_name,
      rows,
      headAtCollection
    );
    return {
      evidence: this.evidenceStore.forNode(node.id),
      bundleHash,
      source: 'collected',
      collectedAt: Date.now(),
      headAtCollection,
    };
  }

  private fileHeadSha(filePath: string): string | null {
    try {
      const out = execFileSync(
        'git',
        ['log', '-1', '--format=%H', '--', filePath],
        {
          cwd: this.repoRoot,
          encoding: 'utf8',
          stdio: ['ignore', 'pipe', 'pipe'],
        }
      ).trim();
      return out === '' ? null : out;
    } catch {
      return null;
    }
  }
}

File: src/whygraph/evidence/service.py
from __future__ import annotations

import time
from pathlib import Path
from typing import Callable

from whygraph.backend import SymbolNode
from whygraph.evidence.git import GitEvidenceCollector, collect_git_evidence
from whygraph.evidence.github import (
    GitHubEvidenceCollector,
    collect_github_evidence,
)
from whygraph.evidence.store import EvidenceStore
from whygraph.evidence.types import CollectionResult


class EvidenceService:
    def __init__(
        self,
        store: EvidenceStore,
        git: GitEvidenceCollector,
        github: GitHubEvidenceCollector | None,
        repo_root: Path,
        ttl_seconds: int,
        *,
        now: Callable[[], int] | None = None,
        head_sha_fn: Callable[[str], str | None] | None = None,
    ) -> None:
        self._store = store
        self._git = git
        self._github = github
        self._repo_root = repo_root
        self._ttl_seconds = ttl_seconds
        self._now = now or (lambda: int(time.time()))
        self._head_sha_fn = head_sha_fn or (
            lambda file_path: git.file_head_sha(file_path)
        )

    def for_node(
        self, node: SymbolNode, *, force: bool = False
    ) -> CollectionResult:
        if not force:
            cached = self._check_cache(node)
            if cached is not None:
                return cached
        return self._collect(node)

    def _check_cache(self, node: SymbolNode) -> CollectionResult | None:
        meta = self._store.bundle_meta_for(node.id)
        if meta is None:
            return None
        age = self._now() - meta.built_at
        if age > self._ttl_seconds:
            return None
        if meta.head_at_collection is not None:
            current = self._head_sha_fn(node.file_path)
            if current is None or current != meta.head_at_collection:
                return None
        # head_at_collection is None → trust TTL alone (no git history at
        # collection time means no per-file sha to compare; recollecting on
        # every call would be wasteful).
        return CollectionResult(
            evidence=self._store.for_node(node.id),
            bundle_hash=meta.bundle_hash,
            source="cache",
            collected_at=meta.built_at,
            head_at_collection=meta.head_at_collection,
        )

    def _collect(self, node: SymbolNode) -> CollectionResult:
        git_rows = collect_git_evidence(self._git, node)
        gh_rows = (
            collect_github_evidence(self._github, git_rows)
            if self._github is not None and self._github.is_available()
            else []
        )
        rows = git_rows + gh_rows
        head = self._head_sha_fn(node.file_path)
        now = self._now()
        bundle_hash = self._store.replace(
            node.id, node.qualified_name, rows, head, now=now
        )
        return CollectionResult(
            evidence=self._store.for_node(node.id),
            bundle_hash=bundle_hash,
            source="collected",
            collected_at=now,
            head_at_collection=head,
        )
